pkg/util/deeptest/statictesthelper.go

```diff
@@ -0,0 +1,142 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package deeptest
+
+import (
+	"fmt"
+	"reflect"
+	"regexp"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+)
+
+type staticTestHelper struct {
+	ignorePath []regexp.Regexp
+}
+
+// NewStaticTestHelper creates a new helper to help testing the deep equality of two objects.
+func NewStaticTestHelper(ignorePath []string) *staticTestHelper {
+	pathRegexp := make([]regexp.Regexp, 0, len(ignorePath))
+	for _, path := range ignorePath {
+		pathRegexp = append(pathRegexp, *regexp.MustCompile(path))
+	}
+	return &staticTestHelper{ignorePath: pathRegexp}
+}
+
+func (h *staticTestHelper) shouldIgnore(path string) bool {
+	for _, reg := range h.ignorePath {
+		if reg.MatchString(path) {
+			return true
+		}
+	}
+	return false
+}
+
+// AssertNotEmpty asserts that the value and its every fields are not empty.
+func (h *staticTestHelper) AssertNotEmpty(t *testing.T, valObj reflect.Value, path string) {
+	if h.shouldIgnore(path) {
+		return
+	}
+
+	require.False(t, valObj.IsZero(), path+" should not be empty")
+
+	switch valObj.Type().Kind() {
+	case reflect.Struct:
+		for i := 0; i < valObj.NumField(); i++ {
+			h.AssertNotEmpty(t, valObj.Field(i), path+"."+valObj.Type().Field(i).Name)
+		}
+	case reflect.Ptr:
+		h.AssertNotEmpty(t, valObj.Elem(), path)
+	case reflect.Slice, reflect.Array:
+		require.NotEqual(t, 0, valObj.Len(), path+" should not be empty")
+		for i := 0; i < valObj.Len(); i++ {
+			h.AssertNotEmpty(t, valObj.Index(i), path+fmt.Sprintf("[%d]", i))
+		}
+	case reflect.Bool:
+		require.Equal(t, true, valObj.Bool(), path+" should be true")
+	case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64:
+		require.NotEqual(t, 0, valObj.Int(), path+" should not be 0")
+	case reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64:
+		require.NotEqual(t, 0, valObj.Uint(), path+" should not be 0")
+	case reflect.Float32, reflect.Float64:
+		require.NotEqual(t, 0.0, valObj.Float(), path+" should not be 0")
+	case reflect.String:
+		require.NotEqual(t, "", valObj.String(), path+" should not be empty")
+	case reflect.Map:
+		require.NotEqual(t, 0, valObj.Len(), path+" should not be empty")
+		for _, key := range valObj.MapKeys() {
+			h.AssertNotEmpty(t, valObj.MapIndex(key), path+fmt.Sprintf("[%v]", key))
+		}
+	case reflect.Interface:
+		h.AssertNotEmpty(t, valObj.Elem(), path)
+	case reflect.Func:
+		return
+	default:
+		require.Fail(t, "unsupported type", path)
+	}
+}
+
+// Equal tells whether a and b are deeply equal.
+func (h *staticTestHelper) Equal(t *testing.T, valA, valB reflect.Value, path string) {
```

Comment:
Could we use generic to make it easier to use? For example:

```go
func (h * staticTestHelper[T]) AssertEqual(t *testing.T, obj *T) {
	val := reflect.ValueOf(objA)
        var empty T
	valEmpty := reflect.ValueOf(&empty)
	....
}

func (h * staticTestHelper[T]) AssertEveryFieldNotEqual(t *testing.T, objA, objB *T) {
	valA := reflect.ValueOf(objA)
	valB := reflect.ValueOf(objB)
	....
}

func (h * staticTestHelper[T]) AssertNotEmpty(t *testing.T, objA, objB *T) {
	valA := reflect.ValueOf(objA)
	valB := reflect.ValueOf(objB)
	....
}
```

I think we can also remove the last argument `path` because it seems the outside code always use `$` as an input.

You can introduce a private method `equal(t *testing.T, objA, objB *T, path string)` to be used it internally.

Reply:
Also, we can just use a function instead of an helper object to do the assert:

```
func AssertEqual(objA, objB T, opts ...Option) {
	...
}

func WithIgnorePath(path ...string) Option {
	...
}

// We can use it like this
AssertEqual(t, ctx1, ctx2, WithIgnorePath(path1, path2))
```

Reply:
Yes. But I still kept the helper for internal use to share the `ignorePath` recursively. The option is introduced to change the internal configuration of the helper.